### DISCUSSION TITLE
Use grid-engine for MMORPG

### DISCUSSION THREAD

**Discussion by sbernardotti** [Annoraaq/grid-engine#210] [General]
Hi, I'm trying to make a MMORPG like Argentum Online with Phaser and Colyseus. Fortunately, I found this library, but I don't know if it can be adapted to work as I expect. These are my doubts:

- Can grid-engine delegate the collisions stuff to the server?
- How can I move a Character when a server command arrives?
- How can I sync characters' movement and their position when, eg, tab is in background or lag/pin is high?

Thank you!

**Comment by Annoraaq** [comment]
Hi @sbernardotti 

Great to hear you plan to use GridEngine for your project. Let me try to answer your questions.

### How can I move a Character when a server command arrives?

Simply call the `move()` or `moveTo()` methods of GridEngine when the command arrives. In the example code these methods are invoked when the player presses the arrow keys but you can call them whenever you want.

### How can I sync characters' movement and their position when, eg, tab is in background or lag/pin is high?

You can use `setPosition()` to enforce a position for a character.

I think here it depends a lot on the concrete implementation whether this is working smoothly or not. If you just update the GridEngine positions every 'x' milliseconds with the data from the server it could mean that player positions suddenly change and players 'jump' to new positions. This behavior can be observed in many online games when players have a laggy internet connection. At a first look I can only think of two strategies (and probably many combinations of those) for handling a lag:

1. block the game
2. let the game run and correct the positions after the connection is established again

Which one to use depends a lot on your requirements.

### Can grid-engine delegate the collisions stuff to the server?

While there is no direct functionality for delegating this information to the server I see some possible solutions:

1.  You could turn off collisions in GridEngine and handle it yourself. For example before every movement you could consult the server and ask if the desired position is blocked or not. And if it is blocked you simply don't allow a movement command to GridEngine. I guess you have to put some more thoughts into the details for this to address performance issues but it could be a point to start.

2. You can also let GridEngine handle collisions but character positions being handled by the server. So that means that you need to sync the character positions regularly with those of the server. (see above)

------------

I never used it for an online game but it is definitely something I personally would want GridEngine to be capable of. So if you decide to go with GridEngine and hit a wall, open up a discussion or an issue and it might end up as a new feature for GridEngine.

I hope this helps a bit.

**Comment by lendsp** [comment]
I was doing on vacation

I used nodejs, mongodb, socket.io

I actually opted to turn off collisions and control it on the server, but it is possible to maintain and have a double check

whenever a player moves it is best to teleport to the point of origin and move to the destination

at the moment my game is not online but there are some videos on youtube

https://www.youtube.com/watch?v=6LRUotM0ubc

**Reply by Annoraaq** [reply]
That's pretty cool. Please keep us updated about it. I'm very interested in how well GridEngine is suited for an MMO.

**Reply by lendsp** [reply]
> That's pretty cool. Please keep us updated about it. I'm very interested in how well GridEngine is suited for an MMO.

you can play https://obeli.herokuapp.com/ also my client side is javascript and is open for now

**Comment by sbernardotti** [comment]
Hi everyone!

Im calling moveTo method when a message arrives from server and it woks fine (partially...)

What I need is that every character in every client takes the same path to get to the target position, but I couldn't make it. I tried modifying [MoveToConfig props](https://annoraaq.github.io/grid-engine/typedoc/interfaces/MoveToConfig.html) with no success.
I want it to move exactly like the sender client does (which is the simplest movement/path)

I'm attaching a video showing the issue (Im playing on right tab)

[movement.webm](https://user-images.githubusercontent.com/26163877/179370950-74b99db3-835e-4596-b329-04ef060f1f7b.webm)

Thank you!

**Reply by Annoraaq** [reply]
Do I understand it correctly that the right window (the sending client) is controlled by you via keyboard and `move` method and the server then sends `moveTo` commands to the other client (left window)?

I could imagine a couple of reasons for that behavior:

1. The clients have a different state when calling `moveTo`. That would lead to different outcomes and it can be non-trivial in a server/client architecture to keep the state synchronized.

2. `moveTo` and `move` do not behave equivalent. That would mean that `moveTo` to a neighboring field does not take the shortest path and would therefore be a bug in pathfinding.

3. `moveTo` is non-deterministic. That should not be the case. Maybe when different browsers use different order to store items in Sets or Maps. But I don't think that is the case here and I would only look into that if the other two reasons can be excluded.

I would probably do the following experiment:
In your local client try not to use `move` but use `moveTo` instead and see what happens. That way you could find out whether `moveTo` and `move` behave differently.

**Reply by Annoraaq** [reply]
A quick update: I did the mentioned experiment and experienced some unwanted behavior:
If a position is blocked  and I try to move there via `moveTo`, the player should just STOP if it is blocked. But instead the player walks around the tile. This seems to be a bug and maybe it is causing your problem, too. I will keep you updated.

Edit:
I found that this is a bug in the pathfinding algorithm (Bfs works, but Bidirectional Search produces the error).

**Reply by sbernardotti** [reply]
Thank you! I'll replace move() with moveTo() in the sender client.

 I'll waiting for news with that bug :)

**Reply by lendsp** [reply]
> Hi everyone!
> 
> Im calling moveTo method when a message arrives from server and it woks fine (partially...)
> 
> What I need is that every character in every client takes the same path to get to the target position, but I couldn't make it. I tried modifying [MoveToConfig props](https://annoraaq.github.io/grid-engine/typedoc/interfaces/MoveToConfig.html) with no success. I want it to move exactly like the sender client does (which is the simplest movement/path)
> 
> I'm attaching a video showing the issue (Im playing on right tab)
> 
>  movement.webm 
> Thank you!

if moveto not successful you can use setPosition before MoveTo to ever sync, step by step... i did it in mobs but you need calculate all

**Reply by Annoraaq** [reply]
Two possibly related bugs #301 and #302 were closed with today's release v2.18.1.
@sbernardotti can you check if this solves your problem?
